nex-rs/src/server/server_trait.rs

```diff
@@ -66,31 +62,39 @@ pub trait Server: EventHandler {
 
         self.get_mut_base().socket = Some(socket);
 
-        let clients = Arc::clone(&self.get_base().clients);
+        let clients_lock = Arc::clone(&self.get_base().clients);
         let ping_kick_thread = tokio::spawn(async move {
             let mut invertal = time::interval(Duration::from_secs(3));
             invertal.tick().await;
 
             loop {
                 invertal.tick().await;
-                let mut clients = clients.lock().await;
 
-                for client in clients.iter_mut() {
+                for client in clients_lock
+                    .write()
+                    .await
+                    .iter_mut()
+                    .map(|c| async { c.write().await })
+                {
+                    let mut client = client.await;
                     if let Some(timer) = client.get_kick_timer() {
                         client.set_kick_timer(Some(timer.saturating_sub(3)));
                     }
                 }
 
-                *clients = clients
-                    .iter()
-                    .filter_map(|c| {
-                        if c.get_kick_timer() == Some(0) {
-                            None
-                        } else {
-                            Some(c.clone())
+                let mut to_remove = vec![];
+                for (i, client) in clients_lock.read().await.iter().enumerate() {
+                    if let Some(time) = client.read().await.get_kick_timer() {
+                        if time == 0 {
+                            to_remove.push(i);
                         }
-                    })
-                    .collect::<Vec<ClientConnection>>();
+                    }
+                }
+
+                let mut clients = clients_lock.write().await;
+                for index in to_remove {
+                    clients.remove(index);
+                }
```

Comment:
It feels like allocating a new vector, iterating over two vectors, and reordering items in a vector (during the remove) seems like a lot just for a filter.  Is there another way to handle this?

Reply:
This was the first idea I had. The reason it's this way is we can't have async closures as it's unsafe. Because of this, we can't use normal iter functions nor call into_iter() on the vector as the the values get moved which isn't allowed by the borrow checker. 

I personally am unable to find another solution but this one allows for the best in-the-middle performance. As it iterates over the list to remove it uses a read lock so for a duration of the loop it wont hold a universal lock over the rest of the packets (except for packets from new clients).

Reply:
I noticed the initial read then write locks, and figured that's what the intention was.  I'm not sure if we can say it's the best in-the-middle performance without having actual performance metrics.

This was one of the things I was referring to in the other comment about async, iterators, and whether we have the right structure.

Reply:
If we want to handle the client kicking in another thread, I don't see any other way to handle it. We're going to need to acquire a lot but we also can't call async functions inside a closure unless we map them to futures and await them. We could use the `futures` crate but I believe it inherently does the same thing

Reply:
How does the current method of this look?